### PR TITLE
Store trie nodes in DB

### PR DESCRIPTION
**Pull request by rphmeier**
...only for chain head. might be nice to have some history, but getting pruning right is quite a lot of logic. the alternative is to run archive nodes for now.

### CODE REVIEW

**Comment by rphmeier** [comment]
memoization of storage root:
  - every call to `storage_root` produces a backend-specific "transaction" which is necessary to apply the latest updates. this is memoized along with the produced root.
  - if no changes to storage are made between calls, the calculation is not re-done
  - the DB-based backend produces a "transaction" which is a memory overlay of all trie nodes inserted and deleted
  - the in-memory backend produces a "transaction" of the raw changed storage values
  - after executing each block, the client takes this transaction and applies it.

File: substrate/client/db/src/lib.rs [hunk 1]
```diff
@@ -384,17 +488,14 @@ impl client::backend::Backend for Backend {
...
 				transaction.put(columns::META, meta::BEST_BLOCK, &key);
 			}
-			for (key, val) in operation.pending_state.changes.into_iter() {
-				match val {
-					Some(v) => { transaction.put(columns::STATE, &key, &v); },
-					None => { transaction.delete(columns::STATE, &key); },
+			for (key, (val, rc)) in operation.pending_state.updates.drain() {
+				if rc > 0 {
+					transaction.put(columns::STATE, &key.0[..], &val);
+
```
Comment (gavofyork) [inline substrate/client/db/src/lib.rs]: stray newline

File: polkadot/api/src/lib.rs [hunk 2]
```diff
@@ -280,49 +277,46 @@ impl<S: state_machine::Backend> ClientBlockBuilder<S>
 {
 	// executes a extrinsic, inherent or otherwise, without appending to the list
```
Comment (gavofyork) [inline polkadot/api/src/lib.rs]: gah. bad comment.

Comment (rphmeier) [reply polkadot/api/src/lib.rs]: https://github.com/paritytech/polkadot/blame/513b544fb505b8c59f07adf10e8dc81de34d2a73/polkadot/api/src/lib.rs#L278

:)

(i remember this from the relevant PR but didn't bring it up then for some reason)

**Comment by rphmeier** [comment]
gonna ice this for the moment

**Comment by rphmeier** [comment]
test added

**Pull request merged**